### PR TITLE
Refactoring and new features

### PR DESCRIPTION
**Pull request by kumom**
Features added:
- Build the project for web instead of `nw.js` (although `npm run start:nw` still allows serving the files in `nw.js`)
- Resizable panel (see `Resizer.tsx`)
- Source language selector (see `Toolbar.tsx`)
- Move peg-toolbar to the global level `Toolbar.tsx` and "Definition PEG" will only be shown when it is available for the selected source language.
- Integrate a formatter (prettier) in the building process
- Spinners shown when the panel is loading/rendering
- HypergraphView is not rendered from scratch when an overlay is removed. A method called `removeOverlay` (in `hypergraph.ts`) is added.

Notes:
- I migrated the whole project from `Vue` to `React` because making something non-reactive in Vue is very hacky. The migration only took me 6 hours in total, so if you really prefer Vue, I can always change it back. I do not have personal preferences between two frameworks, but I do think a larger community like React provides more available components and plugins (I used react-monaco-editor in this case).
- Tip for merging: run `npm run format` first to see content-based diffs.

### CODE REVIEW

**Comment by e-avidan** [comment]
@kumom - Wow, so much stuff 😅 
I went over the changes and most of them seem pretty great to me

It's a little hard to see with all the re-formatting, but I noticed that you're attempting to merge into `lang-c` instead of `master`, and there are a few changes that I merged there that will cause conflicts, specifically:
* Changes to types that I moved out of `hypergraph.ts` and `patterns.ts`
* Methods and consts that I moved out of `patterns.ts` and `semantics.ts`

Also, like we discussed there's the point of changing the `Definition PEG` checkbox into buttons (like there are currently in the nw:js version on `master`). I'm even thinking of not doing buttons, instead having a Dropdown in place of the checkbox that's called `Analysis Type`, defaulting to `None` - and with other options such as `Semantics` and `PTA` (each running the respective function currently exists in `main.ts`).
In this scenario, whenever an analysis is selected from the dropdown, it'll `removeOverlay` the previous results and `addOverlay` the new ones (which would be kool)

So I think in summary
* You should rebase from `master` (and change the PR to be towards it as well). I think you shouldn't have changed much in the files I touched so mostly you can just take what I did, add your specfic changes and reformat again
* Change the checkbox into a dropdown
* LMK and I can help you verify that everything's running smoothly :-)

**Comment by corwin-of-amber** [comment]
I think we can do a dropdown *with* checkboxes, because we may want to have multiple analyses run on the same code. Wouldn't that be cool?

**Comment by kumom** [comment]
> I think we can do a dropdown _with_ checkboxes, because we may want to have multiple analyses run on the same code. Wouldn't that be cool?

That's not difficult to implement when we do many more than one analysis to run...but do you think we should implement it right now? That would then be a dropdown with only one checkbox

File: src/analysis/patternSerializer.ts [hunk 1]
```diff
@@ -18,7 +17,9 @@ const KNOWN_FUNCTIONS = {
 
 export function loadPattern(patternString: string): RoutePatternDefinition {
   const routeSplit = patternString.split("||");
-  assert(routeSplit.length === 2);
+  if (routeSplit.length !== 2) {
+    throw "routeSplit.length !== 2";
+  }
```
Comment (e-avidan) [inline src/analysis/patternSerializer.ts:22]: why did you change the assertions into throws?

File: src/analysis/patternSerializer.ts [hunk 2]
```diff
@@ -110,7 +113,12 @@ function* _serializeDefinition(def: PatternDefinition) {
   } = def;
 
   if (typeof labelPred === "function") {
-    assert(KNOWN_FUNCTIONS[labelPred.name]);
+    if (
+      KNOWN_FUNCTIONS[labelPred.name] === null ||
+      KNOWN_FUNCTIONS[labelPred.name] === undefined
```
Comment (e-avidan) [inline src/analysis/patternSerializer.ts:118]: could just be `!KNOWN_FUNCTIONS[labelPred.name]`

File: src/analysis/syntax.ts [hunk 3]
```diff
@@ -1,4 +1,34 @@
-import * as SetOps from "../infra/setops";
+import * as SetOps from "infra/setops";
+
+export type TypeScriptSyntaxToken =
```
Comment (e-avidan) [inline src/analysis/syntax.ts:3]: these were actually on purpose in `syntax.d.ts` (which is a typescript definitions file) - any reason you reverted that change?

File: src/syntax/parser.ts [hunk 4]
```diff
@@ -5,37 +5,51 @@ class SkippingLexer implements nearley.Lexer {
...
+  constructor(lexer: moo.Lexer) {
     this.lexer = lexer;
     this.skip = new Set(["WS", "COMMENT"]);
   }
 
   next() {
     do {
       const token = this.lexer.next();
-      if (!(token && this.skip.has(token.type))) {
+      if (!(token != null && this.skip.has(token.type!))) {
```
Comment (e-avidan) [inline src/syntax/parser.ts:16]: why `!=`? 
It's better to use `!==`, but the original is cleaner IMO

Comment (e-avidan) [reply src/syntax/parser.ts:16]: Also is `token.type!` valid syntax?

**Review by e-avidan — CHANGES_REQUESTED** [review]
Cool! 
Glad to see we don't have any more conflicts - left a few comments.

Also, regarding PTA - I think it should just be as simple as calling `performPointsToAnalysis` (like you called `seman` which is set to `semanticAnalysis_TS `)

File: src/syntax/parser.ts (hunk 4 above)
Comment (kumom) [reply src/syntax/parser.ts:16]: This changes were made to the original code, because `tsconfig.json` throws error on this line. But it's fine now. I didn't go line by line during this merging, so I missed this line.

File: src/analysis/syntax.ts (hunk 3 above)
Comment (kumom) [reply src/analysis/syntax.ts:3]: Because I set `"isolatedModules": true` in `tsconfig.json` and `syntax.d.ts` is global. See [this post](https://stackoverflow.com/questions/42233987/how-to-configure-custom-global-interfaces-d-ts-files-for-typescript) for why I made this change. You can change it back later by adapting `tsconfig.json`, but I am not sure writing definition files in such a way is recommended, since we are not writing a library and the type files are not grouped under a folder.

File: src/analysis/patternSerializer.ts (hunk 2 above)
Comment (kumom) [reply src/analysis/patternSerializer.ts:118]: Yeah you are right

File: src/analysis/patternSerializer.ts (hunk 1 above)
Comment (kumom) [reply src/analysis/patternSerializer.ts:22]: Because `assert` is not a native module and I don't wanna use `console.assert` because we have not decided our final environment. You can use `npm run build` to see the error it throws.

**Comment by kumom** [comment]
> Cool!
> Glad to see we don't have any more conflicts - left a few comments.
> 
> Also, regarding PTA - I think it should just be as simple as calling `performPointsToAnalysis` (like you called `seman` which is set to `semanticAnalysis_TS `)

Regarding PTA, I know producing the analysis graph is simple as that. However, the UI needs to adapt to this. Otherwise, where should we see the result? That's why I want to know how your `SourceNavigator` looks like and wait to integrate it until we decide on how to change the UI.

File: src/analysis/patternSerializer.ts (hunk 1 above)
Comment (e-avidan) [reply src/analysis/patternSerializer.ts:22]: maybe let's install `assert` as a dependency?

Comment (kumom) [reply src/analysis/patternSerializer.ts:22]: Yeah I have installed it now.

Comment (kumom) [reply src/analysis/patternSerializer.ts:22]: `assert` is a CommonJS module, so it does not really fit in automatically. I have switched to `console.assert` for now. Let me know if you have a better idea.

**Comment by corwin-of-amber** [comment]
🎉
